plugins/utils/todo-comments.nix

```
@@ -1,290 +1,429 @@
-          max_line_len = cfg.highlight.maxLineLen;
-        };
-        inherit (cfg) colors;
-        search = {
-          inherit (cfg.search) command args;
-          pattern = helpers.ifNonNull' cfg.search.pattern (
-            if isList cfg.search.pattern then
-              (map helpers.mkRaw cfg.search.pattern)
-            else
-              helpers.mkRaw "[[${cfg.search.pattern}]]"
-          );
-        };
-      } // cfg.extraOptions;
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-      extraPackages = [ cfg.ripgrepPackage ];
-      extraConfigLua = ''
-        require("todo-comments").setup${helpers.toLuaObject setupOptions}
-      '';
+  settingsExample = {
+    highlight = {
+      pattern = [
+        ".*<(KEYWORDS)\s*:"
+        ".*<(KEYWORDS)\s*"
+      ];
+    };
+  };
 
-      keymaps = flatten (
-        mapAttrsToList (
+  extraOptions = {
+    keymaps =
+      let
+        mkKeymapOption =
           optionName: funcName:
-          let
-            keymap = cfg.keymaps.${optionName};
-
-            cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
-            keywords = optionalString (keymap.keywords != null) " keywords=${keymap.keywords}";
-          in
-          optional (keymap != null) {
-            mode = "n";
-            inherit (keymap) key options;
-            action = ":${funcName}${cwd}${keywords}<CR>";
-          }
-        ) commands
-      );
-
-      # Automatically enable plugins if keymaps have been set
-      plugins = mkMerge [
-        (mkIf (cfg.keymaps.todoTrouble != null) { trouble.enable = true; })
-        (mkIf (cfg.keymaps.todoTelescope != null) { telescope.enable = true; })
-      ];
+          mkCompositeOption "Keymap settings for the `:${funcName}` function." {
+            key = mkOption {
+              type = types.str;
+              default = null;
+              description = "Key for the `${funcName}` function.";
+            };
+
+            cwd = mkOption {
+              type = types.nullOr types.str;
+              description = "Specify the directory to search for comments";
+              default = null;
+              example = "~/projects/foobar";
+            };
+
+            keywords = mkOption {
+              type = types.nullOr types.str;
+              description = ''
+                Comma separated list of keywords to filter results by.
+                Keywords are case-sensitive.
+              '';
+              default = null;
+              example = "TODO,FIX";
+            };
+
+            options = keymaps.mapConfigOptions;
+          };
+      in
+      mapAttrs mkKeymapOption keymapsActions;
+
+    ripgrepPackage = mkPackageOption {
+      name = "ripgrep";
+      default = pkgs.ripgrep;
     };
+  };
+
+  extraConfig = cfg: {
+    assertions = [
+      {
+        assertion =
+          cfg.keymaps.todoTelescope != null && cfg.keymaps.todoTelescope.key != null
```

Instead of null-checking each step, we can use the fallback operator:

```suggestion
          cfg.keymaps.todoTelescope.key or null != null
```

---

Thanks, I was hoping this could be simplified with some kind of null conditional operator.